src/ui/model_viewer_widget.cc

```diff
@@ -108,84 +105,76 @@ void BuildImageModel(const Image& image, const Camera& camera,
       Eigen::Vector4f(-image_width, -image_height, focal_length, 1);
 
   // Image plane as two triangles.
+  triangle_data.reserve(triangle_data.size() + 2);
```

Comment:
Remove this, as it will lead to extremely O(N^2) allocation cost.

Reply:
I might be wrong on this but I think it isn't O(N^2). We're reserving the total required size at the beginning of ModelViewerWidget::UploadImageData and ModelViewerWidget::UploadMovieGrabberData, which means that the reserve here is a noop. But I agree that this ends up being unnecessary in this case, if we assume that BuildImageModel will only be called from these places.

Reply:
And even if someone ends up using BuildImageModel without having reserved the total required size at the beginning, it probably still makes sense to keep it here because otherwise the allocation would happen during emplace back which isn't any better I guess.

Reply:
If we ever use this function somewhere else then this will end up as O(N^2). If the emplace_back() detects that the vector is out of capacity it will just increase the vector by a certain factor.

Reply:
Ah I see, you're right of course, I'll fix that tonight when I get home.